### ISSUE TITLE
URL is immutable from SQLAlchemy 1.4

### ISSUE THREAD

**Issue by failys** [pallets/flask-sqlalchemy#885]
### Expected Behavior

Where the SQLAlchemy object is created using the 'second possibility' documented in the SQLAlchemy code, and an app context exists, calling the create_all method in the SQLAlchemy class should be successful.

### Actual Behavior

Because URL objects are [immutable from SQLAlchemy 1.4](https://docs.sqlalchemy.org/en/14/changelog/migration_14.html#the-url-object-is-now-immutable) calling the get_engine method in the _EngineConnector class will cause an AttributeError exception to be thrown due to an attempt to manually modify a previously created URL in the apply_driver_hacks method in the SQLAlchemy class, i.e.

```python
if sa_url.drivername.startswith("mysql"):
            sa_url.query.setdefault("charset", "utf8")
```

I haven't had the chance to test this, but it seems you could replace

```python
  sa_url.query.setdefault("charset", "utf8")
```

with

```python
  sa_url.update_query_dict({"charset" : "utf8"})
```

### Environment

* Python version: 3.8
* Flask-SQLAlchemy version:  2.4.4
* SQLAlchemy version: 1.4.0b1

**Comment by arabidopsis** [comment]
It's likely to be more complicated than that. The sa_url will return a new updated url that will have to
be propagated back to the _EngineConnector that called the driver hacks so it can then give the updated url
to create_engine.

**Comment by mcr** [comment]
I also see this on line 914, where it does:
~~~~
            # if it's not an in memory database we make the path absolute.
            if not detected_in_memory:
                sa_url.database = os.path.join(app.root_path, sa_url.database)
~~~~

as a simple work around, sa_url.database does not need to be updated if it is already absolute.

**Comment by Rukeith** [comment]
I got the same issue while my SQLalchemy is `==1.3.23`

**Comment by mdantonio** [comment]
Version 1.4 released yesterday and it is now automatically installed as dependency, so current flask-sqlalchemy no longer works on fresh installations (but can be easily fixed by forcing the installation of SQLAlchemy 1.3.23)

I open issue #928 because I did not see this other issue, please consider to close the duplicate

**Comment by davidism** [comment]
Locking to avoid more "me too" comments. We are aware of the issue and it will be fixed.

**Comment by davidism** [comment]
fixed by #933, will release 2.5 soon

**Comment by Saibamen** [comment]
For people who doesn't want to update anything - fix for this is just to add:
```
SQLAlchemy<1.4
```

in your `requirements.txt` file

via https://stackoverflow.com/questions/65610809/retrieving-data-from-rds-gives-attributeerror-sqlalchemy-cimmutabledict-immuta
